dropshot/src/websocket.rs

```diff
@@ -0,0 +1,410 @@
+// Copyright 2022 Oxide Computer Company
+/*!
+ * Implements websocket upgrades as an Extractor for use in API route handler
+ * parameters to indicate that the given endpoint is meant to be upgraded to
+ * a websocket.
+ *
+ * This exposes a [`HyperUpgraded`] to a user-provided async future,
+ * which will be spawned to handle the incoming connection.
+ */
+
+use crate::api_description::ExtensionMode;
+use crate::{
+    ApiEndpointBodyContentType, Extractor, ExtractorMetadata, HttpError,
+    RequestContext, ServerContext,
+};
+use async_trait::async_trait;
+use http::{header, Response, StatusCode};
+use hyper::upgrade::OnUpgrade;
+use hyper::{upgrade, Body};
+use schemars::JsonSchema;
+use serde_json::json;
+use sha1::{Digest, Sha1};
+use slog::Logger;
+use std::future::Future;
+use std::sync::Arc;
+use tokio::task::JoinHandle;
+
+// re-export such that consumers need not version-match with us in their Cargo.toml
+pub use hyper::upgrade::Upgraded as HyperUpgraded;
+
+/**
+ * WebsocketUpgrade is an Extractor used to upgrade and handle an HTTP request
+ * as a websocket when present in a Dropshot endpoint's function arguments.
+ *
+ * The consumer of this *must* call [WebsocketUpgrade::handle] (or if they do
+ * not wish to upgrade a given connection, [WebsocketUpgrade::do_not_upgrade])
+ * on the owned instance in the endpoint's fn, or else WebsocketUpgrade's
+ * implementation of [Drop::drop] will panic.
```

Comment:
Why's that?  (why not implicitly call `do_not_upgrade()` on drop, say?)

It looks like it's a non-fatal error to handle the upgrade more than once (we produce an error) but it's fatal to not handle it at all.  I don't have a strong idea here but I just wonder if there's a way to make it harder to accidentally panic in production.  (One thought I had was that `WebsocketUpgrade` could be parametrized by an impl of some trait that would allow us to immediately and automatically invoke it exactly once for you...but then I'm not sure what the body of the function would do.  Also, I'm not sure if it's desirable to be able to do stuff or look at the request before the upgrade -- presumably it is (e.g., authentication).  Then I thought this could be a different kind of endpoint...but then I worry I'm overthinking it.)

Reply:
for the interface to be more discoverable (because `.handle()` *has* to be called for any of this to work; this way nobody's wondering why their connection isn't getting upgraded, even if they weren't originally clear on the docs) and less prone to incorrect use (e.g. accidental inclusion of a WebsocketUpgrade in a param list due to copy/paste errors, causing us to eventually generate an confusing/incorrect client from the resulting schema)

if we're very concerned about panics making it into production, we can make it be an error log in Drop instead, but that's less likely to be caught at development time -- maybe panic if debug assertions are on, else log?